File: tests/src/org.sonatype.sisu/sisu-inject-plexus/1.4.2/src/test/java/org_sonatype_sisu/sisu_inject_plexus/PlexusXmlBeanConverterTest.java
/*
 * Copyright and related rights waived via CC0
 *
 * You should have received a copy of the CC0 legalcode along with this
 * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
 */
package org_sonatype_sisu.sisu_inject_plexus;

import static org.assertj.core.api.Assertions.assertThat;

import com.google.inject.Guice;
import com.google.inject.TypeLiteral;
import java.util.Properties;
import org.junit.jupiter.api.Test;
import org.sonatype.guice.plexus.converters.PlexusXmlBeanConverter;

public class PlexusXmlBeanConverterTest {
    @Test
    void convertsXmlElementsToArray() {
        PlexusXmlBeanConverter converter = newConverter();

        String[] values = (String[]) converter.convert(TypeLiteral.get(String[].class), """
                <values>
                    <value>alpha</value>
                    <value>beta</value>
                </values>
                """);

        assertThat(values).containsExactly("alpha", "beta");
    }

    @Test
    void loadsImplementationFromThreadContextClassLoader() {
        PlexusXmlBeanConverter converter = newConverter();
        ClassLoader originalClassLoader = Thread.currentThread().getContextClassLoader();
        Thread.currentThread().setContextClassLoader(PlexusXmlBeanConverterTest.class.getClassLoader());
        try {
            EmptyRole role = (EmptyRole) converter.convert(TypeLiteral.get(EmptyRole.class), """
                    <role implementation="%s"> </role>
                    """.formatted(TcclLoadedRole.class.getName()));

            assertThat(role).isInstanceOf(TcclLoadedRole.class);
        } finally {
            Thread.currentThread().setContextClassLoader(originalClassLoader);
        }
    }

    @Test
    void fallsBackToPeerClassLoaderForImplementation() {
        PlexusXmlBeanConverter converter = newConverter();
        ClassLoader originalClassLoader = Thread.currentThread().getContextClassLoader();
        Thread.currentThread().setContextClassLoader(new RejectingClassLoader(originalClassLoader));
        try {
            EmptyRole role = (EmptyRole) converter.convert(TypeLiteral.get(EmptyRole.class), """
                    <role implementation="%s"> </role>
                    """.formatted(PeerLoadedRole.class.getName()));

            assertThat(role).isInstanceOf(PeerLoadedRole.class);
        } finally {
            Thread.currentThread().setContextClassLoader(originalClassLoader);
        }
    }

    @Test
    void fallsBackToClassForNameForBootstrapDefaultImplementation() {
        PlexusXmlBeanConverter converter = newConverter();
        ClassLoader originalClassLoader = Thread.currentThread().getContextClassLoader();
        Thread.currentThread().setContextClassLoader(null);
        try {
            Properties properties = (Properties) converter.convert(TypeLiteral.get(Properties.class), """
                    <properties implementation="java.util.Properties">
                        <property>
                            <name>server</name>
                            <value>native-image</value>
                        </property>
                    </properties>
                    """);

            assertThat(properties).containsEntry("server", "native-image");
        } finally {
            Thread.currentThread().setContextClassLoader(originalClassLoader);
        }
    }

    @Test
    void convertsPlainTextWithStringConstructor() {
        PlexusXmlBeanConverter converter = newConverter();

        ConstructedValue value = (ConstructedValue) converter.convert(
                TypeLiteral.get(ConstructedValue.class), "from text");

        assertThat(value.value()).isEqualTo("from text");
    }

    private static PlexusXmlBeanConverter newConverter() {
        return Guice.createInjector().getInstance(PlexusXmlBeanConverter.class);
    }

    public static class EmptyRole {
        public EmptyRole() {
        }
    }

    public static final class TcclLoadedRole extends EmptyRole {
        public TcclLoadedRole() {
        }
    }

    public static final class PeerLoadedRole extends EmptyRole {
        public PeerLoadedRole() {
        }
    }

    public static final class ConstructedValue {
        private final String value;

        public ConstructedValue(final String value) {
            this.value = value;
        }

        public String value() {
            return value;
        }
    }

    private static final class RejectingClassLoader extends ClassLoader {
        RejectingClassLoader(final ClassLoader parent) {
            super(parent);
        }

        @Override
        public Class<?> loadClass(final String name) throws ClassNotFoundException {
            throw new ClassNotFoundException(name);
        }
    }
}

File: tests/src/org.jetbrains.kotlin/kotlin-main-kts/2.3.21/src/test/kotlin/org_jetbrains_kotlin/kotlin_main_kts/PlexusXmlBeanConverterTest.kt
/*
 * Copyright and related rights waived via CC0
 *
 * You should have received a copy of the CC0 legalcode along with this
 * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
 */
package org_jetbrains_kotlin.kotlin_main_kts

import org.assertj.core.api.Assertions.assertThat
import org.jetbrains.kotlin.com.google.inject.AbstractModule
import org.jetbrains.kotlin.com.google.inject.Guice
import org.jetbrains.kotlin.com.google.inject.TypeLiteral
import org.jetbrains.kotlin.org.eclipse.sisu.plexus.PlexusBeanConverter
import org.jetbrains.kotlin.org.eclipse.sisu.plexus.PlexusXmlBeanConverter
import org.junit.jupiter.api.Test

public class PlexusXmlBeanConverterTest {
    private val converter: PlexusBeanConverter = Guice.createInjector(
        object : AbstractModule() {
            override fun configure(): Unit {
                bind(PlexusBeanConverter::class.java).to(PlexusXmlBeanConverter::class.java)
            }
        },
    ).getInstance(PlexusBeanConverter::class.java)

    @Test
    public fun convertsXmlCollectionsToArrays(): Unit {
        val values: Array<String> = converter.convert(
            TypeLiteral.get(Array<String>::class.java),
            """
            <values>
              <value>alpha</value>
              <value>beta</value>
            </values>
            """.trimIndent(),
        )

        assertThat(values).containsExactly("alpha", "beta")
    }

    @Test
    public fun convertsTextUsingPublicStringConstructor(): Unit {
        val value: PlexusXmlBeanConverterTextValue = converter.convert(
            TypeLiteral.get(PlexusXmlBeanConverterTextValue::class.java),
            "constructed from text",
        )

        assertThat(value.text).isEqualTo("constructed from text")
    }

    @Test
    public fun loadsXmlImplementationWithThreadContextClassLoader(): Unit {
        val previousClassLoader: ClassLoader? = Thread.currentThread().contextClassLoader
        Thread.currentThread().contextClassLoader = PlexusXmlBeanConverterImplementedBean::class.java.classLoader
        try {
            val bean: PlexusXmlBeanConverterBaseBean = converter.convert(
                TypeLiteral.get(PlexusXmlBeanConverterBaseBean::class.java),
                implementationXml(PlexusXmlBeanConverterImplementedBean::class.java.name, "tccl"),
            )

            assertThat(bean).isInstanceOf(PlexusXmlBeanConverterImplementedBean::class.java)
            assertThat(bean.name).isEqualTo("tccl")
        } finally {
            Thread.currentThread().contextClassLoader = previousClassLoader
        }
    }

    @Test
    public fun fallsBackToRoleClassLoaderWhenThreadContextClassLoaderCannotLoadImplementation(): Unit {
        withRejectingThreadContextClassLoader {
            val bean: PlexusXmlBeanConverterBaseBean = converter.convert(
                TypeLiteral.get(PlexusXmlBeanConverterBaseBean::class.java),
                implementationXml(PlexusXmlBeanConverterImplementedBean::class.java.name, "role loader"),
            )

            assertThat(bean).isInstanceOf(PlexusXmlBeanConverterImplementedBean::class.java)
            assertThat(bean.name).isEqualTo("role loader")
        }
    }

    @Test
    public fun fallsBackToClassForNameWhenRoleClassUsesBootstrapClassLoader(): Unit {
        withRejectingThreadContextClassLoader {
            val number: Number = converter.convert(
                TypeLiteral.get(Number::class.java),
                """
                <number implementation="java.lang.Integer">42</number>
                """.trimIndent(),
            )

            assertThat(number).isEqualTo(42)
        }
    }

    private fun implementationXml(implementation: String, name: String): String =
        """
        <bean implementation="$implementation">
          <name>$name</name>
        </bean>
        """.trimIndent()

    private fun withRejectingThreadContextClassLoader(block: () -> Unit): Unit {
        val thread: Thread = Thread.currentThread()
        val previousClassLoader: ClassLoader? = thread.contextClassLoader
        thread.contextClassLoader = RejectingClassLoader()
        try {
            block()
        } finally {
            thread.contextClassLoader = previousClassLoader
        }
    }

    public class RejectingClassLoader : ClassLoader(null) {
        override fun loadClass(name: String): Class<*> {
            throw ClassNotFoundException(name)
        }
    }
}

public class PlexusXmlBeanConverterTextValue(public val text: String)

public open class PlexusXmlBeanConverterBaseBean {
    public var name: String = ""
}

public class PlexusXmlBeanConverterImplementedBean : PlexusXmlBeanConverterBaseBean()
